File: src/BusinessTemplates/CSModernBusinessApp/OpenSilverBusinessApplication.Server/Services/AuthenticationService.cs
using Microsoft.AspNetCore.Identity;
using OpenRiaServices.Server;
using OpenRiaServices.Server.Authentication;
using $ext_safeprojectname$.Models;

namespace $ext_safeprojectname$.Services;

[EnableClientAccess]
public class AuthenticationService : DomainService, IAuthentication<User>
{
    public User Login(string userName, string password, bool isPersistent, string customData)
    {
        var signInManager = ServiceContext.GetRequiredService<SignInManager<User>>();

        var response = signInManager.PasswordSignInAsync(userName, password, isPersistent, lockoutOnFailure: false)
            .GetAwaiter().GetResult();

        if (response.Succeeded)
        {
            return GetAuthenticatedUser(userName);
        }
        else if (response.IsLockedOut)
        {
            throw new NotImplementedException();
        }
        else if (response.RequiresTwoFactor)
        {
            // customdata can be used to pass 2fa token
            throw new NotImplementedException();
        }

        return null;
    }

    private User GetAnonymousUser() => new() { Name = string.Empty, Roles = [] };

    private User GetAuthenticatedUser(string userName)
    {
        var userManager = ServiceContext.GetRequiredService<UserManager<User>>();

        var user = userManager.FindByNameAsync(userName)
            .GetAwaiter().GetResult();

        if (user == null) return null;

        return new User()
        {
            Name = user.UserName,
            Roles = user.Roles,
            PasswordAnswer = user.PasswordAnswer,
            PasswordQuestion = user.PasswordQuestion
        };
    }

    public User Logout()
    {
        var signInManager = ServiceContext.GetRequiredService<SignInManager<User>>();

        signInManager.SignOutAsync()
            .GetAwaiter().GetResult();

        return GetAnonymousUser();
    }

    public User GetUser()
    {
        var identity = ServiceContext.GetRequiredService<IHttpContextAccessor>().HttpContext.User.Identity;

        return identity?.IsAuthenticated == true
            ? GetAuthenticatedUser(identity?.Name)
            : GetAnonymousUser();
    }

    public void UpdateUser(User user)
    {
        throw new NotImplementedException();
    }
}

File: src/BusinessTemplates/VBModernBusinessApp/OpenSilverBusinessApplication.Server/Services/AuthenticationService.vb
Imports Microsoft.AspNetCore.Http
Imports Microsoft.AspNetCore.Identity
Imports Microsoft.Extensions.DependencyInjection
Imports OpenRiaServices.Server
Imports OpenRiaServices.Server.Authentication
Imports $ext_safeprojectname$.Models

Namespace Services

    ''' <summary>
    ''' Domain Service responsible for authenticating users when they log on to the application.
    ''' </summary>
    <EnableClientAccess>
    Public Class AuthenticationService
        Inherits DomainService
        Implements IAuthentication(Of User)

        Public Function Login(userName As String, password As String, isPersistent As Boolean, customData As String) As User Implements IAuthentication(Of User).Login

            Dim signInManager = ServiceContext.GetRequiredService(Of SignInManager(Of User))()

            Dim response = signInManager.PasswordSignInAsync(userName, password, isPersistent, False).GetAwaiter().GetResult()

            If response.Succeeded Then
                Return GetAuthenticatedUser(userName)
            End If

            Return Nothing
        End Function

        Private Function GetAnonymousUser() As User
            Return New User With {.Name = String.Empty, .Roles = Array.Empty(Of String)}
        End Function

        Private Function GetAuthenticatedUser(userName As String) As User

            Dim userManager = ServiceContext.GetRequiredService(Of UserManager(Of User))()

            Dim user = userManager.FindByNameAsync(userName).GetAwaiter().GetResult()

            If user Is Nothing Then
                Return Nothing
            End If

            Return New User() With
        {
            .Name = user.UserName,
            .Roles = user.Roles,
            .PasswordAnswer = user.PasswordAnswer,
            .PasswordQuestion = user.PasswordQuestion
        }

        End Function

        <Query(IsComposable:=False)>
        Public Function GetUser() As User Implements IAuthentication(Of User).GetUser

            Dim identity = ServiceContext.GetRequiredService(Of IHttpContextAccessor)().HttpContext.User.Identity

            If identity.IsAuthenticated Then
                Return GetAuthenticatedUser(identity.Name)
            Else
                Return GetAnonymousUser()
            End If
            
        End Function

        Public Function Logout() As User Implements IAuthentication(Of User).Logout

            Dim SignInManager = ServiceContext.GetRequiredService(Of SignInManager(Of User))()

            SignInManager.SignOutAsync().GetAwaiter().GetResult()

            Return GetAnonymousUser()

        End Function

        Public Sub UpdateUser(user As User) Implements IAuthentication(Of User).UpdateUser
        End Sub

    End Class

End Namespace